File: backend/internal/modules/user/user_service.go
package user

import (
	"citramascoweb-backend/internal/dto"
	"citramascoweb-backend/internal/modules/auth"
	"errors"
	"strings"
	"time"

	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

type userService struct {
	userRepo UserRepositoryInterface
}

func NewUserService(userRepository UserRepositoryInterface) *userService {
	return &userService{userRepo: userRepository}
}

func (s *userService) GetAllCustomer() ([]auth.User, error) {

	users, err := s.userRepo.GetAllCustomer()
	if err != nil {
		return nil, err
	}

	return users, nil

}

func (s *userService) GetAllByRole(role string) ([]auth.User, error) {
	users, err := s.userRepo.GetAllByRole(role)
	if err != nil {
		return nil, err
	}

	return users, nil
}

func (s *userService) Delete(id string) error {
	err := s.userRepo.Delete(id)
	if err != nil {
		return err
	}

	return nil
}

func (s *userService) CreateUser(request *dto.CreateUserRequest) (*auth.User, error) {
	if len(request.Password) < 8 {
		return nil, errors.New("password must be at least 8 characters long")
	}
	if strings.Contains(request.Password, " ") {
		return nil, errors.New("password must not contain spaces")
	}

	exists, err := s.userRepo.CheckUsername(request.Username)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, errors.New("username or email is already taken")
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(request.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	now := time.Now()
	user := &auth.User{
		Id:        uuid.New().String(),
		FirstName: request.FirstName,
		LastName:  request.LastName,
		Username:  request.Username,
		Password:  string(hash),
		Phone:     request.Phone,
		Email:     request.Email,
		Address:   request.Address,
		Role:      auth.Role(request.Role),
		CreatedAt: now,
		UpdatedAt: now,
	}

	err = s.userRepo.Create(user)
	if err != nil {
		return nil, err
	}

	return user, nil
}

func (s *userService) UpdateRole(id string, role auth.Role) error {
	err := s.userRepo.UpdateRole(id, role)
	if err != nil {
		return err
	}

	return nil
}

File: frontend/src/services/admin/userService.ts
import api from '../api'
import type { ApiResponse, User, UserRole } from '@/types'

export const userService = {
  /**
   * GET /api/customers
   * List of customers (role = user)
   */
  getCustomers: async (): Promise<User[]> => {
    const response = await api.get<ApiResponse<User[]>>('/customers')
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to fetch customers')
    }
    return response.data.data
  },

  /**
   * GET /api/users/?role={role}
   * List of users by role
   */
  getByRole: async (role: UserRole): Promise<User[]> => {
    const response = await api.get<ApiResponse<User[]>>('/users/', {
      params: { role },
    })
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to fetch users')
    }
    return response.data.data
  },

  /**
   * POST /api/users
   * Create a new user with an assigned role
   */
  create: async (data: {
    first_name: string
    last_name: string
    username: string
    password: string
    phone: string
    email: string
    address?: string
    role: UserRole
  }): Promise<User> => {
    const response = await api.post<ApiResponse<User>>('/users/', data)
    if (!response.data.success || !response.data.data) {
      throw new Error(response.data.message || 'Failed to create user')
    }
    return response.data.data
  },

  /**
   * PATCH /api/users/:id/role
   * Assign/update a user's role
   */
  updateRole: async (id: string, role: UserRole): Promise<string> => {
    const response = await api.patch<ApiResponse<{ message: string }>>(`/users/${id}/role`, {
      role,
    })
    if (!response.data.success) {
      throw new Error(response.data.message || 'Failed to update user role')
    }
    return response.data.message
  },

  /**
   * DELETE /api/users/:id
   */
  delete: async (id: string): Promise<string> => {
    const response = await api.delete<ApiResponse<{ message: string }>>(`/users/${id}`)
    if (!response.data.success) {
      throw new Error(response.data.message || 'Failed to delete user')
    }
    return response.data.message
  },
}
